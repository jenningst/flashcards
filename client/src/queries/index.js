import { gql } from 'apollo-boost';

const GET_PACKS = gql`
  query fetchPacks {
    fetchPacks {
      _id
      name
    }
  }
`;

const GET_FLASHCARDS_BY_PACK = gql`
  query fetchFlashcardsByPack($id: String!) {
    fetchFlashcardsByPack(id: $id) {
      _id
      text
      answer
      user_id
      pack_id
    }
  }
`;

const CREATE_PACK = gql`
  mutation createPack($input: CreatePackInput!) {
    createPack(input: $input) {
      pack {
        _id
        name
      }
      details {
        code
        success
        message
      }
    }
  }
`;

const CREATE_FLASHCARD = gql`
  mutation createFlashcard($input: CreateFlashcardInput!) {
    createFlashcard(input: $input) {
      card {
        _id
        text
        answer
        user_id
        pack_id
      }
      details {
        code
        success
        message
      }
    }
  }
`;

export {
  GET_PACKS,
  CREATE_PACK,
  CREATE_FLASHCARD,
  GET_FLASHCARDS_BY_PACK
};