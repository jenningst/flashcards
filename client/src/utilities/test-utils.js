const uuidv4 = require('uuid/v4');

function seedMockPack(pack, n = 1) {
  const filter = uuidv4(); // generate a filter
  const name = uuidv4(); // generate a name
  const modes = ['REVIEW_MODE', 'WRITE_MODE']; // TODO: remove hard-coding
  const mode = modes[getRandomInt(0, 1)]; // generate a random mode
  let numToSeed = getRandomInt(1, n || pack.length - 1); // generate a random number of cards
  
  let cards = [];
  for (let i = 0; i < numToSeed; i++) {
    let card = getRandomElement(pack, pack.length - 1);
    card.id = uuidv4();
    cards.push(card);
  }
 
  return { mode, name, filter, cards };
};

function getRandomElement(arr, ceiling) {
  return arr[getRandomInt(0, ceiling)];
};

function getRandomInt(min, max) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min)) + min;
};

export { seedMockPack };