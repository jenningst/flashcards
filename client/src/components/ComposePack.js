import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Mutation } from 'react-apollo';
import { CREATE_PACK, GET_PACKS } from '../queries';
import { ReactComponent as Back } from '../components/icons/svg/back.svg';
import { Link, withRouter } from 'react-router-dom';
import { MediumButton } from '../components/elements/Button';
import { Title2, Caption3 } from './elements/Text';

function ComposePack({ history }) {
  const [packName, setPackName] = useState('');
  const handleNameChange = (e) => setPackName(e.target.value);

  return (
    <ComposePackWrapper className="ComposePack">
      <Header className="ComposePack__header">
        {/* <IconButton
          className="ComposePack__button-back"
          data-testid="button-back"
          to="/"
        /> */}
        <ButtonGroup className="button-label-group">
          <Link to="/">
            <BackIcon
              className="ComposePack__button-back"
              data-testid="button-back"
            />
          </Link>
          <Caption3>Dashboard</Caption3>
        </ButtonGroup>
      </Header>
      <MainSection>

        <Form className="ComposePack__form form">
          <Title2 className="form__title">Create a Pack</Title2>
          <div className="input-group">
            <label
              className="form__label"
              htmlFor="pack-name"
            >
              Pack Name
            </label>
            <input
              className="form__input"
              id="pack-name"
              name="pack-name"
              value={packName}
              onChange={(e) => handleNameChange(e)}
              maxLength="30"
              />
          </div>
          
          <Mutation
            mutation={CREATE_PACK}
            update={(cache, { data }) => {
              // get our current packs from cache
              const { fetchPacks } = cache.readQuery({ query: GET_PACKS });
              // write back to the cache
              cache.writeQuery({
                query: GET_PACKS,
                data: { fetchPacks: [...fetchPacks, data.createPack.pack] },
              });
            }}
          >
            {(addPack) => (
              <FormFooter>
                <StyledSubmit
                  className="form__button-submit"
                  type="submit"
                  disabled={packName ? false : true}
                  onClick={e => {
                    if (packName !== '') {
                      addPack({ variables: { input: {
                        name: packName
                      }}});
                      history.push("/"); // go home
                    }
                  }}
                >
                  Submit
                </StyledSubmit>
              </FormFooter>
            )}
          </Mutation>

        </Form>
      </MainSection>
    </ComposePackWrapper>
  );
};

const ComposePackWrapper = styled.div`
  box-sizing: border-box;
  display: grid;
  grid-template-rows: minmax(11%, 13%) 1fr;
  grid-template-areas:
    "header"
    "form";
  height: 100%;
  background: ${props => props.theme.color.main.offWhite};
  color: ${props => props.theme.color.fonts.charleston};
`;

const BackIcon = styled(Back)`
  height: 2rem;
  width: 2rem;
`;

const ButtonGroup = styled.div`
  display: flex;
  flex-flow: row nowrap;
  justify-content: flex-start;
  align-items: center;

  svg {
    margin-right: 1rem;
  }

  h6 {
    color: ${props => props.theme.color.main.primary};
  }
`;

const Header = styled.header`
  grid-area: header;
  padding: 1.5rem;
`;

const MainSection = styled.section`
  grid-area: form;
  display: flex;
  flex-flow: row nowrap;
  justify-content: center;
  align-items: center;
  padding: 1.5rem;
`;

const Form = styled.form`
  padding: 3rem 1rem 2rem 1rem;
  background: ${props => props.theme.color.main.pureWhite};
  border-radius: 1rem;
  box-shadow: 0px 10px 18px -11px rgba(120,119,120,1);

  .form__title {
    margin-bottom: 2rem;
    font-weight: 500;
    color: ${props => props.theme.color.main.primary};
  }
  
  .form__label {
    font-family: 'Rubik', sans-serif;
    font-size: 15px;
    font-weight: 400;
    line-height: 20px;
    margin: 0;
    color: ${props => props.theme.color.fonts.grey};
  }
  .input-group {
    display: flex;
    flex-flow: column nowrap;
    justify-content: flex-start;
    align-items: flex-start;
    width: 100%;
    margin-bottom: 2rem;

    .form__input {
      height: 2rem;
      margin-top: .50rem;
      padding: .25rem 1rem .25rem 1rem;
      font-size: 1rem;
      border: 1px solid ${props => props.theme.color.borders.lightGrey};
      border-radius: .30rem;

      &:focus {
        outline: 2px solid ${props => props.theme.color.main.primaryHover};
      }
    }
  }
`;

const FormFooter = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

const StyledSubmit = styled(MediumButton)`
  max-width: 20rem;
  padding-left: 2rem;
  padding-right: 2rem;
  background: ${props => props.theme.color.main.primary};
  color: ${props => props.theme.color.fonts.pureWhite};
  border-radius: .50rem;
  box-shadow: 0px 10px 18px -11px rgba(120,119,120,1);

  &:disabled {
    background: ${props => props.theme.color.main.secondary};
    box-shadow: none;
  }
  &:hover {
    background-color: ${props => props.theme.color.main.primaryHover};
  }
`;

ComposePack.propTypes = {
  history: PropTypes.object,
}

export default withRouter(ComposePack);
