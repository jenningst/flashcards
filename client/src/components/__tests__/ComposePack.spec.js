import React from 'react';
// import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { MockedProvider } from 'react-apollo/test-utils';

import {
  cleanup,
  fireEvent,
  render,
} from '@testing-library/react';

import ComposePack from '../ComposePack';
import { CREATE_PACK } from '../../queries/index';

// TODO: mock out the gql provider
const mocks = [
  {
    request: {
      query: CREATE_PACK,
      variables: {
        name: 'React Testing',
      },
    },
    result: {
      data: {
        card: { _id: '1', name: 'React Testing' },
      },
    },
  },
];

afterEach(cleanup);

const renderComponent = () => 
  render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <MemoryRouter initialEntries={['/']}>
        <ComposePack />
      </MemoryRouter>
    </MockedProvider>
  );

describe('<ComposePack /> spec', () => {
  it.todo('assert component matches snapshot');

  it('assert page greeting is rendered', () => {
    const { getByText } = renderComponent();
    expect(getByText(/^create a pack$/i)).toBeInTheDocument();
  });

  it('assert new pack name input is rendered with default value', () => {
    const { getByLabelText } = renderComponent();
    const inputElement = getByLabelText(/^new pack name:$/i);
    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toBeEmpty();
  });

  it('assert a submit button is rendered', () => {
    const { getByText } = renderComponent();
    expect(getByText(/^submit$/i)).toBeInTheDocument();
  });

  it('assert submit button is initially disabled', () => {
    const { getByText } = renderComponent();
    expect(getByText(/^submit$/i)).toBeDisabled();
  });

  it('assert submit button is enabled upon valid input', async () => {
    const newValue = 'What do you call a fake noodle?';
    const mockEvent = { target: { value: newValue}}
    const { getByLabelText, findByText } = renderComponent();
    
    fireEvent.change(getByLabelText(/^new pack name:$/i), mockEvent);
    expect(await findByText(/^submit$/i)).toBeEnabled();
  });

  it.todo('assert clicking on submit button calls gql mutation');

  it.todo('assert clicking on submit button routes');
});