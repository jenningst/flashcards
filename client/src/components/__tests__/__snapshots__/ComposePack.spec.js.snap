// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ComposePack /> spec assert component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ComposePack sc-iwsKbI dborcn"
  >
    <header
      class="ComposePack__header sc-gqjmRU fVwKRL"
    >
      <a
        href="/"
        style="display: flex; height: 2rem; width: 2rem;"
      >
        <svg
          class="ComposePack__button-back sc-gZMcBi dcrjLy"
          style="height: 100%; width: 100%; overflow: hidden;"
        >
          back.svg
        </svg>
      </a>
    </header>
    <section
      class="ComposePack__form form sc-VigVT FaJHk"
    >
      <h1
        class="form__title sc-bdVaJa beESmM"
      >
        Create a Pack
      </h1>
      <label
        for="pack-input-name"
      >
        Pack Name: 
      </label>
      <input
        class="form__input"
        id="pack-input-name"
        placeholder="Enter a Pack Name"
        value=""
      />
      <button
        class="form__button-submit"
        disabled=""
        type="submit"
      >
        Submit
      </button>
    </section>
  </div>
</DocumentFragment>
`;
